src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,151 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class Http2Connection
+    {
+        private int InitialStreamWindowSize { get; } = 65535;
+
+        private class Http2StreamWindowManager
+        {
+            // See comment on ConnectionWindowThreshold.
+            protected int _streamWindowUpdateRatio;
+            protected int _delivered;
+            protected int _streamWindowSize;
+
+            protected readonly Http2Connection _connection;
+            protected readonly Http2Stream _stream;
+
+            public Http2StreamWindowManager(Http2Connection connection, Http2Stream stream)
+            {
+                _connection = connection;
+
+                _stream = stream;
+                _streamWindowSize = connection.InitialStreamWindowSize;
+                _streamWindowUpdateRatio = _connection._pool.Settings._streamWindowUpdateRatio;
+                _stream.Trace($"StreamWindowSize: {StreamWindowSize}, StreamWindowThreshold: {StreamWindowThreshold}, streamWindowUpdateRatio: {_streamWindowUpdateRatio}");
+            }
+
+            internal int StreamWindowSize => _streamWindowSize;
+
+            internal int StreamWindowThreshold => _streamWindowSize / _streamWindowUpdateRatio;
+
+            public virtual void AdjustWindow(int bytesConsumed)
+            {
+                Debug.Assert(bytesConsumed > 0);
+                Debug.Assert(_delivered < StreamWindowThreshold);
+
+                if (!_stream.ExpectResponseData)
+                {
+                    // We are not expecting any more data (because we've either completed or aborted).
+                    // So no need to send any more WINDOW_UPDATEs.
+                    return;
+                }
+
+                _delivered += bytesConsumed;
+                if (_delivered < StreamWindowThreshold)
+                {
+                    return;
+                }
+
+                int windowUpdateSize = _delivered;
+                _delivered = 0;
+
+                Task sendWindowUpdateTask = _connection.SendWindowUpdateAsync(_stream.StreamId, windowUpdateSize);
+                _connection.LogExceptions(sendWindowUpdateTask);
+            }
+        }
+
+        private class DynamicHttp2StreamWindowManager : Http2StreamWindowManager
+        {
+            private DateTime _lastWindowUpdate = DateTime.Now;
```

Comment:
`DateTime` is probably not the best option to measure time differences. I remember reading in the past that it has granularity of 16ms. Also `DateTime.Now` will suffer during daylight saving time changes since it depends on the current machine locale.
I know that we use `Environment.TickCount64` for example here:
https://github.com/dotnet/runtime/blob/6ab1cee00ba41b308009c88864aba46b0f61f061/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1982-L1982

Reply:
>  remember reading in the past that it has granularity of 16ms

I think this is only true for OS timers (`Timer` `Thread.Sleep`, `Task.Delay`), not `DateTime.Now`. In my experience it's much more accurate, otherwise it wouldn't make sense to expose [`DateTime.Ticks`](https://docs.microsoft.com/en-us/dotnet/api/system.datetime.ticks?view=net-5.0#remarks), that said, I need to check what native call does it use.

`Environment.TickCount(64)` is one of my favorite candidates in the competition of bad naming, it doesn't give ticks, but milliseconds. We need microsecond accuracy here.

Reply:
https://docs.microsoft.com/en-us/archive/blogs/ericlippert/precision-and-accuracy-of-datetime
>On my machine this says “same” eight or nine times, and then suddenly the Ticks property jumps by about 160000, which is 16 milliseconds, a 64th of a second. (Different flavours of Windows might give you different results, depending on details of their thread timing algorithms and other implementation details.)

Though it's >10 years old, so I'm not sure what the current state of things is.

Reply:
Good point, IDK if it has improved or not, but I will check how the code works with Stopwatch.

Reply:
DateTime is accurate, but Stopwatch can be even better (performance counter on Windows) and is better suited for looking at time intervals.

For DateTime, `UtcNow` is better since `Now` has to adjust for timezones.